<a href="https://colab.research.google.com/github/rodrigoreisbr-tech/data_science/blob/master/Introdu%C3%A7%C3%A3o_ao_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introdução ao Pandas

Pandas é uma biblioteca do Python, provavelmente a mais popular de todas quando se trata de *Data Science*.

Por meio do Pandas, você consegue importar dados (arquivos `csv`e `xls`, por exemplo), tratar esses dados, transformá-los e realizar análises completas dos mesmos.

Uma vez que você importa um conjunto de dados usando Pandas, fica muito fácil fazer coisas do tipo:

* Extrai informações estatísticas
  * Qual a média, mediana, valores máximos e mínimos?
  * Qual é a distribuição das suas variáveis?
  * Qual a correlação entre duas variáveis quaisquer?
* Exportar os dados para um novo formato de arquivo
* Visualizar gráficos dos mais diferentes tipos
* Alimentar modelos de *machine learning* feitos em cima do Scikit-learn

Pandas é construído em cima de outra biblioteca extremamente popular, o **NumPy**. Quem já utilizou esta, vai encontrar muita similaridade com aquela.

## Instalando e utilizando o Pandas

Caso você esteja rodando o código na sua máquina local, o Pandas pode ser instalado pela linha de comando usando o seu gerenciador de pacotes (`pip` ou `conda`).

Dependendo de qual deles seja, o comando que você deve usar é

`pip install pandas`

ou 

`conda install pandas`

Como estamos usando o Colab neste curso, o Pandas vem instalado por padrão. Ou seja, a única coisa que você precisa é importar o pacote. Normalmente, para importar qualquer pacote basta apenas usar o comando `import nome_do_pacote`.

No entanto, é muito comum a gente importar o Pandas usando `import pandas as pd` para abreviar o nome da biblioteca. Se você reparar em projetos de outros cientistas de dados, vai ver que esse é (provavelmente) a maneira que eles utilizam.

In [0]:
# importar a biblioteca pandas
import pandas as pd

### Importar arquivos `csv`

Para dar um exemplo de como é simples importar dados com o Pandas, veja o seguinte exemplo.

Pelo site [Yahoo Finance](https://finance.yahoo.com/quote/BBAS3.SA/history?p=BBAS3.SA), baixei os dados da ação ordinária do Banco do Brasil (BBAS3) em formato `csv` e disponibilizei o arquivo [neste link](https://raw.githubusercontent.com/rodrigoreisbr-tech/data_science/master/dataset/BVSP.csv). 

Importar esse mesmo arquivo usando o Pandas é tão simples quanto executar a função `pd.read_csv("local_do_arquivo.csv")`, informando qual o endereço que o `csv` se encontra (endereço na internet ou caminho na máquia local.



In [0]:
# importar o arquivo csv para o Pandas
df = pd.read_csv("https://raw.githubusercontent.com/rodrigoreisbr-tech/data_science/master/dataset/BVSP.csv")

Pronto! O arquivo `BVSP.csv` foi importado com sucesso e já está pronto para ser usado neste *notebook*. 

### Conhecendo os componentes básicos do Pandas

Os dois componentes básicos que a gente deve conhecer quando lidando com Pandas são `Series` e `DataFrame`. 

<center><img src="https://raw.githubusercontent.com/carlosfab/curso_data_science_na_pratica/master/modulo_02/componentes_pandas.png"></center>

Simplificadamente, você pode pensar o `DataFrame` como sendo uma planilha de Excel, e `Series` como sendo apenas uma coluna individual.

Apesar de parecer conceitualmente simples, estas duas estruturas nativas do Pandas facilitam muito o trabalho com dados, uma vez que elas podem armazenar qualquer tipo de dado.

* `type()` - mostra qual o tipo da variável


In [0]:
# ver o tipo da variável df
type(df)

pandas.core.frame.DataFrame

In [0]:
# ver o tipo de uma coluna da variável df
type(df['Date'])

pandas.core.series.Series

### Ver dimensões do DataFrame

Veja novamente a imagem acima. Basicamente, o arquivo que importamos se parece com uma simples tabela de Excel, composta por linhas e colunas. Para você ver o tamanho dessa "tabela", o que significa ver o formato (*shape*) dela, basta executar `df.shape`.

In [0]:
# ver o tamanho do dataframe (formato)
df.shape

(249, 7)

Quando você executar a célula acima, vai receber como *output* os valores `(249, 7)`. Isso significa que o arquivo importado possui 249 linhas e 7 colunas. 

### Conhecendo os dados

Uma vez que você importou a sua base de dados para o Pandas, existem muitos atributos e métodos nativos da estrutura *DataFrame* que facilitam muito a exploração de dados.

Uma das principais funções da biblioteca, e que você irá usar em praticamente todos os seus projetos é `df.head()` e `df.tail()`.

O arquivo `csv` com os dados da ação BBAS3 contém 248 linhas, mas o normal é você lidar com milhares ou centenas de milhares de linhas. Obviamente, seria inviável se tivessemos que olhar cada linha para entender como os dados estão apresentados.

Na verdade, quando a gente importa um *dataset*, queremos dar uma olhadinha rápida em algumas entradas, só para ter noção dos dados que iremos lidar. Isso é feito facilmente com:

* `df.head()` - exibe as 5 primeiras entradas do conjunto de dados
* `df.tail()` - exibe as 5 últimas entradas do conjunto de dados

In [0]:
# mostrar as 5 primeiras entradas do DataFrame
df.head()

Date     Open     High      Low    Close  Adj Close     Volume
0  2019-06-10  97828.0  97856.0  96782.0  97467.0    97467.0  3584000.0
1  2019-06-11  97467.0  98985.0  97467.0  98960.0    98960.0  4922400.0
2  2019-06-12  98960.0  99240.0  97831.0  98321.0    98321.0  4948800.0
3  2019-06-13  98321.0  99364.0  98321.0  98774.0    98774.0  5594000.0
4  2019-06-14  98776.0  98981.0  97601.0  98040.0    98040.0  9184900.0

In [0]:
# mostrar as 5 últimas entradas do DataFrame
df.tail()

Date         Open  ...     Adj Close      Volume
244  2020-06-03  91048.00000  ...  93002.000000  15137900.0
245  2020-06-04  92993.00000  ...  93829.000000  11488000.0
246  2020-06-05  93839.00000  ...  94637.000000  13320300.0
247  2020-06-08  94640.00000  ...  97645.000000  11096500.0
248  2020-06-09  97644.34375  ...  96746.546875         0.0

[5 rows x 7 columns]

Um dos motivos da popularidade do Pandas é por causa dessa capacidade de conseguir mostrar os dados como se estivessem em uma tabela, um formato bem amigável para a compreensão do nosso cérebro.

Compare com aquela imagem lá em cima, do Excel. É um formato que estamos bem habituados a enxergar.

Vamos supor que você precise extrair apenas os nomes das colunas do seu *DataFrame* - basta executar:

In [0]:
# ver os nomes das colunas
df.columns

Index(['Date', 'Open', 'High', 'Low', 'Close', 'Adj Close', 'Volume'], dtype='object')

Conhecer o tipo de variável que está representado em cada coluna é essencial. Por exemplo:

* Quando temos uma coluna que trata de receita, despesa ou lucro, é desejável que lidemos com variáveis do tipo `float`.
* Quando estamos lidando com anos (2017, 2018, 2019), iremos desejar trabalhar com variáveis do tipo `int`.
* Quanto temos datas completas (`2019-12-30 07:37`), iremos desejar usar o formato `datetime`, para conseguir manipular adequadamente o *dataset*.

Para conhecer os tipos de variáveis de cada coluna, use `df.dtypes`.

In [0]:
df.dtypes

Date          object
Open         float64
High         float64
Low          float64
Close        float64
Adj Close    float64
Volume       float64
dtype: object

Uma outra maneira de descobrir o tipo das variáveis e ainda ver os valores ausentes nas células, é por meio do método `df.info()`.

Veja como ele consegue retornar essas informações de maneira tabular.

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 249 entries, 0 to 248
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       249 non-null    object 
 1   Open       248 non-null    float64
 2   High       248 non-null    float64
 3   Low        248 non-null    float64
 4   Close      248 non-null    float64
 5   Adj Close  248 non-null    float64
 6   Volume     248 non-null    float64
dtypes: float64(6), object(1)
memory usage: 13.7+ KB


### Selecionando colunas

Há diversas maneiras de selecionar um sub-conjunto de dados em uma estrutura *DataFrame*.

Na maioria dos casos, o que queremos fazer é selecionar apenas uma coluna de todo o *DataFrame*.

Para selecionar uma variável, você deve colocar o nome da coluna entre colchetes, referenciando a variável onde os seus dados foram importados.

In [0]:
# selecionar a variável (coluna) "High" da nossa variável df
df["High"]

0      97856.00000
1      98985.00000
2      99240.00000
3      99364.00000
4      98981.00000
          ...     
244    93710.00000
245    94132.00000
246    97356.00000
247    97693.00000
248    97644.34375
Name: High, Length: 249, dtype: float64

Caso o nome das coluna não contenha espaços em branco e caracteres especiais, você consegue selecionar ela de outra maneira (muito mais prática, na minha opinião).

No lugar de `df["High"]` você pode usar `df.High`, que terá o mesmo efeito.

In [0]:
# selecionar a variável (coluna) "High" da nossa variável df
df.High

0      97856.00000
1      98985.00000
2      99240.00000
3      99364.00000
4      98981.00000
          ...     
244    93710.00000
245    94132.00000
246    97356.00000
247    97693.00000
248    97644.34375
Name: High, Length: 249, dtype: float64

### Calculando a média de uma coluna

Quando executamos o código `df.High`, você viu acima que foi impressa apenas a coluna de valores `High`. Para encontrar a média desses valores, basta utilizar `df.High.mean()`.

In [0]:
# calcular a média da coluna "High"
df.High.mean()

101078.76348286291

Isso significa que a média dos valores da coluna `High` é `46.57639677327938`. Como vamos ver ao longo do curso, existem muitas outras medidas importantes.

## Trabalhando com datas (`datetime`)

Quando você importar um arquivo que tenha datas, provavelmente o Pandas não irá reconhecer as mesmas de maneira automática.

Quando você checou `df.info()`, viu que a coluna `Date` estava como `object`. No caso, ela é apenas texto (`string`) e não permite a extração de muita coisa útil.

Vamos converter a coluna usando `pd.to_datetime` e informando qual o formato da nossa data.

In [0]:
# antes
df.Date

0      2019-06-10
1      2019-06-11
2      2019-06-12
3      2019-06-13
4      2019-06-14
          ...    
244    2020-06-03
245    2020-06-04
246    2020-06-05
247    2020-06-08
248    2020-06-09
Name: Date, Length: 249, dtype: object

In [0]:
# converter coluna Date em datetime
df.Date = pd.to_datetime(df.Date, format="%Y-%m-%d")

In [0]:
# depois
df.Date

0     2019-06-10
1     2019-06-11
2     2019-06-12
3     2019-06-13
4     2019-06-14
         ...    
244   2020-06-03
245   2020-06-04
246   2020-06-05
247   2020-06-08
248   2020-06-09
Name: Date, Length: 249, dtype: datetime64[ns]

Olhe como agora eu consigo acessar apenas os componentes de uma data, individualmente, usando `df.Date.dt`

In [0]:
df.Date.dt.year

0      2019
1      2019
2      2019
3      2019
4      2019
       ... 
244    2020
245    2020
246    2020
247    2020
248    2020
Name: Date, Length: 249, dtype: int64

In [0]:
df.Date.dt.month

0      6
1      6
2      6
3      6
4      6
      ..
244    6
245    6
246    6
247    6
248    6
Name: Date, Length: 249, dtype: int64

In [0]:
df.Date.dt.day

0      10
1      11
2      12
3      13
4      14
       ..
244     3
245     4
246     5
247     8
248     9
Name: Date, Length: 249, dtype: int64


No entanto, há vários formatos que um usuário pode lançar como sendo datas. Você pode encontrar exemplos como:

* 31/06/15
* 31/06/2015
* 31-6-15
* 2015-6-31
* 31 de junho de 2015

Para lidar com tudo isso, você vai ter que entender a documentação. Seguem dois links que serão úteis para isso:

* https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html?highlight=to_datetime#pandas.to_datetime
* https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior